### PR TITLE
Add support for bot integrations.

### PR DESCRIPTION
**Pull request by XuaTheGrate**
## Summary

Adds support for bot integrations. I'm not entirely confident about the design, so feedback welcome.

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [x] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/integrations.py [hunk 1]
```diff
@@ -79,18 +114,20 @@ class Integration:
         The behaviour of expiring subscribers. Aliased to ``expire_behavior`` as well.
     expire_grace_period: :class:`int`
         The grace period (in days) for expiring subscribers.
-    user: :class:`User`
-        The user for the integration.
+    user: Optional[:class:`User`]
+        The user for the integration. In the case of a bot application, this is the person who invited it.
```
Comment (NCPlayz) [inline discord/integrations.py:118]:
```suggestion
        The user for the integration. In the case of a bot application, this is the person who invited it.
        
        .. versionadded:: 2.0
```

File: discord/integrations.py [hunk 2]
```diff
@@ -79,18 +114,20 @@ class Integration:
...
     account: :class:`IntegrationAccount`
         The integration account information.
     synced_at: :class:`datetime.datetime`
         When the integration was last synced.
+    application: Optional[:class:`Application`]
+        The application this integration belongs to. Can be ``None`` if it is not a bot integration.
```
Comment (NCPlayz) [inline discord/integrations.py:124]:
```suggestion
        The application this integration belongs to. Can be ``None`` if it is not a bot integration.
        
        .. versionadded:: 2.0
```

**Review by NCPlayz — CHANGES_REQUESTED** [review]
Noticed missing `versionadded` on the two new `Integration` attributes

File: discord/http.py [hunk 3]
```diff
@@ -745,7 +745,7 @@ def edit_custom_emoji(self, guild_id, emoji_id, *, name, roles=None, reason=None
         return self.request(r, json=payload, reason=reason)
 
     def get_all_integrations(self, guild_id):
-        r = Route('GET', '/guilds/{guild_id}/integrations', guild_id=guild_id)
+        r = Route('GET', '/guilds/{guild_id}/integrations?include_applications=True', guild_id=guild_id)
```
Comment (NCPlayz) [inline discord/http.py:748]:
```suggestion
        r = Route('GET', '/guilds/{guild_id}/integrations', guild_id=guild_id)
```

applications integrations are provided by default now.

File: discord/integrations.py [hunk 4]
```diff
@@ -103,7 +140,7 @@ def __repr__(self):
     def _from_data(self, integ):
         self.id = _get_as_snowflake(integ, 'id')
         self.name = integ['name']
-        self.type = integ['type']
+        self.type = integration_type = integ['type']
```
Comment (NCPlayz) [inline discord/integrations.py:143]:
```suggestion
        self.type = integ['type']
```

File: discord/integrations.py [hunk 5]
```diff
@@ -114,9 +151,21 @@ def _from_data(self, integ):
...
+        user = integ.get('user')
+        if user is not None:
+            self.user = User(state=self._state, data=user)
+        else:
+            self.user = None
+
         self.account = IntegrationAccount(**integ['account'])
 
+        if application is not None:
+            self.application = Application(**application)
```
Comment (NCPlayz) [inline discord/integrations.py:165]:
```suggestion
            self.application = Application(state=self._state, data=application)
```

File: discord/integrations.py [hunk 6]
```diff
@@ -52,6 +52,41 @@ def __init__(self, **kwargs):
...
+        self.id = _get_as_snowflake(data, 'id')
+        self.name = data['name']
+        self.icon = data['icon']
+        self.description = data['description']
+        self.summary = data['summary']
+
+        self.bot = User(state=state, data=data['bot'])
+
+    def __repr__(self):
+        return '<Application id={0.id} name={0.name!r} bot={0.bot!r}>'.format(self)
```
Comment (NCPlayz) [inline discord/integrations.py:88]:
```suggestion
        return f'<Application id={self.id} name={self.name!r} bot={self.bot!r}>'
```

File: discord/integrations.py [hunk 7]
```diff
@@ -52,6 +52,41 @@ def __init__(self, **kwargs):
...
+
+    def __init__(self, *, state, data):
+        self._state = state
+        self.id = _get_as_snowflake(data, 'id')
+        self.name = data['name']
+        self.icon = data['icon']
+        self.description = data['description']
+        self.summary = data['summary']
+
+        self.bot = User(state=state, data=data['bot'])
```
Comment (NCPlayz) [inline discord/integrations.py:85]:
```suggestion
        bot = data.get('bot')
        if bot is not None:
            self.bot = User(state=self._state, data=bot)
        else:
            self.bot = None
```

File: discord/integrations.py [hunk 8]
```diff
@@ -114,9 +151,21 @@ def _from_data(self, integ):
         self.expire_grace_period = integ['expire_grace_period']
         self.synced_at = parse_time(integ['synced_at'])
 
-        self.user = User(state=self._state, data=integ['user'])
+        application = integ.get('application')
```
Comment (NCPlayz) [inline discord/integrations.py:154]: Maybe move this just before the application check?

**Review by NCPlayz — CHANGES_REQUESTED** [review]

File: discord/integrations.py [hunk 9]
```diff
@@ -231,19 +257,74 @@ async def sync(self) -> None:
...
+        'guild',
+        'id',
+        '_state',
+        'type',
+        'name',
+        'account',
+        'user',
+        'enabled',
+        'application'
+    )
```
Comment (Rapptz) [inline discord/integrations.py:326]:
```suggestion
    __slots__ = Integration.__slots__ + ('application',)
```

**Review by Rapptz — CHANGES_REQUESTED** [review]

File: discord/integrations.py [hunk 10]
```diff
@@ -40,7 +41,9 @@
     from .types.integration import (
         IntegrationAccount as IntegrationAccountPayload,
         Integration as IntegrationPayload,
-    )
+        IntegrationType,
+        IntegrationApplication as IntegrationApplicationPayload
+)
```
Comment (NCPlayz) [inline discord/integrations.py:46]:
```suggestion
    )
```

**Review by NCPlayz — CHANGES_REQUESTED** [review]
Seems like `__repr__` went missing for the classes, and you might also want to change the object creation at `Guild.integrations` to use the new integration classes now.

The new classes should probably also be added to `__all__` and `autoclass`'d in the API Reference.

File: discord/integrations.py [hunk 11]
```diff
@@ -231,19 +288,74 @@ async def sync(self) -> None:
...
-        Forbidden
-            You do not have permission to delete the integration.
-        HTTPException
-            Deleting the integration failed.
-        """
-        await self._state.http.delete_integration(self.guild.id, self.id)
+    __slots__ = ('id', 'name', 'icon', 'description', 'summary', 'user')
+    
+    def __init__(self, *, data: IntegrationApplicationPayload, state):
+        self.id: int = _get_as_snowflake(data, 'id')
```
Comment (Rapptz) [inline discord/integrations.py:315]:
```suggestion
        self.id: int = int(data['id'])
```

**Review by Rapptz — APPROVED** [review]

**Comment by Rapptz** [comment]
I decided to fix these review nits myself. Thanks for the PR.

**Pull request merged**